File: optionals/med-4.0.0/tests/c/2.3.6/test25.c
/*  This file is part of MED.
 *
 *  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
 *  MED is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  MED is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with MED.  If not, see <http://www.gnu.org/licenses/>.
 */


/******************************************************************************
 * - Nom du fichier : test25.c
 *
 * - Description : ecriture de mailles de type MED_POLYEDRE
 *                 dans un maillage MED
 *
 *****************************************************************************/

#include <med.h>
#define MESGERR 1
#include <med_utils.h>

#ifdef DEF_LECT_ECR
#define MODE_ACCES MED_LECTURE_ECRITURE
#elif DEF_LECT_AJOUT
#define MODE_ACCES MED_LECTURE_AJOUT
#else
#define MODE_ACCES MED_CREATION
#endif

int main (int argc, char **argv)


{
  med_idt fid;
  char maa[MED_TAILLE_NOM+1] = "maa1";
  med_int mdim = 3;
  med_int n=2;
  /* connectivite nodale */
  med_int indexp[3] = {1,5,9};
  med_int np = 3;
  med_int indexf[9] = {1,4,7,10,13,16,19,22,25};
  med_int nf = 9;
  med_int conn[24] = {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,
		      20,21,22,23,24};
  /* connectivite descendante */
  med_int indexp2[3] = {1,5,9};
  med_int np2 = 3;
  med_int conn2[8] = {1,2,3,4,5,6,7,8};
  med_int indexf2[8] = {MED_TRIA3,MED_TRIA3,MED_TRIA3,MED_TRIA3,
			MED_TRIA3,MED_TRIA3,MED_TRIA3,MED_TRIA3};
  med_int nf2=8;
  /*                             12345678901234561234567890123456*/
  char nom[MED_TAILLE_PNOM*2+1]="polyedre1       polyedre2       ";
  med_int num[2] = {1,2};
  med_int fam[2] = {0,-1};

  /* Creation du fichier test25.med */
  if ((fid = MEDouvrir("test25.med",MODE_ACCES)) < 0) {
    MESSAGE("Erreur a la creation du fichier test25.med");
    return -1;
  }
  printf("Creation du fichier test25.med \n");
 
  /* Creation du maillage */
  if (MEDmaaCr(fid,maa,mdim,MED_NON_STRUCTURE,
		   "un maillage pour test25") < 0) {
    MESSAGE("Erreur a la creation du maillage");
    return -1;
   }
  printf("Creation du maillage \n");

  /* Ecriture des connectivites des mailles polyedres en mode nodal */
  if (MEDpolyedreConnEcr(fid,maa,indexp,np,indexf,nf,conn,MED_NOD) < 0) {
    MESSAGE("Erreur a l'ecriture de la connectivite des mailles MED_POLYEDRE");
    return -1;
  }
  printf("Ecriture des connectivites de mailles de type MED_POLYEDRE en mode nodal \n"); 

  /* Ecriture des connectivites des mailles polyedres en mode descendant */
  if (MEDpolyedreConnEcr(fid,maa,indexp2,np2,indexf2,nf2,conn2,MED_DESC) < 0) {
    MESSAGE("Erreur a l'ecriture des connectivites des mailles MED_POLYEDRE en mode descendant");
    return -1;
  }
  printf("Ecriture des connectivites de mailles de type MED_POLYEDRE en mode descendant \n"); 

  /* Ecriture des noms des polyedres */
  if (MEDnomEcr(fid,maa,nom,n,MED_MAILLE,MED_POLYEDRE) < 0) {
    MESSAGE("Erreur a l'ecriture des noms des mailles MED_POLYEDRE");
    return -1;
  }
  printf("Ecriture des noms des polyedres \n");

  /* Ecriture des numeros des polyedres */
  if (MEDnumEcr(fid,maa,num,n,MED_MAILLE,MED_POLYEDRE) < 0) {
    MESSAGE("Erreur a l'ecriture des numeros des mailles MED_POLYEDRE");
    return -1;
  }
  printf("Ecriture des numeros des polyedres \n");

  /* Ecriture des numeros des familles des polyedres */
  if (MEDfamEcr(fid,maa,fam,n,MED_MAILLE,MED_POLYEDRE) < 0) {
    MESSAGE("Erreur a l'ecriture des familles des mailles MED_POLYEDRE");
    return -1;
  }
  printf("Ecriture des numeros des familles des polyedres \n");
  
  /* Fermeture du fichier */
  if (MEDfermer(fid) < 0) {
    MESSAGE("Erreur a la fermeture du fichier");
    return -1;
  }
  printf("Fermeture du fichier test25.med \n");

  return 0; 
}

File: optionals/med-4.0.0/tests/f/2.3.6/test25.f
C*  This file is part of MED.
C*
C*  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
C*  MED is free software: you can redistribute it and/or modify
C*  it under the terms of the GNU Lesser General Public License as published by
C*  the Free Software Foundation, either version 3 of the License, or
C*  (at your option) any later version.
C*
C*  MED is distributed in the hope that it will be useful,
C*  but WITHOUT ANY WARRANTY; without even the implied warranty of
C*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
C*  GNU Lesser General Public License for more details.
C*
C*  You should have received a copy of the GNU Lesser General Public License
C*  along with MED.  If not, see <http://www.gnu.org/licenses/>.
C*

C       *******************************************************************************
C       * - Nom du fichier : test25.f
C       *
C       * - Description : ecriture de mailles MED_POLYEDRE dans un maillage MED
C       *
C       ******************************************************************************
	program test25
C       
	implicit none
	include 'med.hf'
C	
	integer*8 fid
	integer cret,mdim
	parameter  (mdim = 3)
	character*32 maa 	
        integer n
	parameter (n=2)
C       Connectivite nodale
	integer np,nf
	parameter (nf=9,np=3)
	integer indexp(np),indexf(nf)
	integer conn(24)
C       Connectivite descendante
	integer np2,nf2
	parameter (nf2=8,np2=3)
	integer indexp2(np2),indexf2(nf2)
	integer conn2(nf2)
	character*16 nom(n)
	integer num(n),fam(n)
C
	data indexp / 1,5,9 /
	data indexf / 1,4,7,10,13,16,19,22,25 /
	data conn / 1,2,3,4,5,6,7,8,9,10,11,12,13,14,
     &              15,16,17,18,19,20,21,22,23,24 /    
	data indexp2 / 1,5,9 /
	data indexf2 / MED_TRIA3,MED_TRIA3,MED_TRIA3,MED_TRIA3,
     &                 MED_TRIA3,MED_TRIA3,MED_TRIA3,MED_TRIA3 /
	data conn2 / 1,2,3,4,5,6,7,8 /
	data nom  / "poly1", "poly2"/ 
	data num  / 1,2 /, fam / 0,-1 /
	data maa /"maa1"/

C       ** Creation du fichier test25.med  **
 	call efouvr(fid,'test25.med',MED_LECTURE_ECRITURE, cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur creation du fichier'
	   call efexit(-1)
	endif      
	print *,'Creation du fichier test25.med'

C       ** Creation du maillage          **
	call efmaac(fid,maa,mdim,MED_NON_STRUCTURE,
     &                 'un maillage pour test25',cret)
	if (cret .ne. 0 ) then
	   print *,'Erreur creation du maillage'
	   call efexit(-1)
	endif      
	print *,cret
	print *,'Creation du maillage'

C       ** Ecriture des connectivites des mailles polyedres en mode nodal **
	call efpece(fid,maa,indexp,np,indexf,nf,conn,MED_NOD,cret) 
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur ecriture connectivite des polyedres'
	   call efexit(-1)
	endif      
	print *,'Ecriture des connectivites des mailles
     & de type MED_POLYEDRE'
	print *,'Description nodale'

C       ** Ecriture des connectivites des mailles polyedres en mode descendant **
	call efpece(fid,maa,indexp2,np2,indexf2,nf2,conn2,MED_DESC,cret) 
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur ecriture connectivite des polyedres'
	   call efexit(-1)
	endif      
	print *,'Ecriture des connectivites des mailles 
     & de type MED_POLYEDRE'
	print *,'Description descendante'

C       ** Ecriture des noms des mailles polyedres          **
	call efnome(fid,maa,nom,n,MED_MAILLE,MED_POLYEDRE,
     &		       cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur ecriture noms des polyedres'
	   call efexit(-1)
	endif      
	print *,'Ecriture des noms des polyedress'

C       ** Ecriture des numeros des mailles polyedres **
	call efnume(fid,maa,num,n,MED_MAILLE,MED_POLYEDRE,
     &                 cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur ecriture numeros des polyedres'
	   call efexit(-1)
	endif      
	print *,'Ecriture des numeros des polyedres'

C	** Ecriture des numeros des familles des segments  **
	call effame(fid,maa,fam,n,
     &              MED_MAILLE,MED_POLYEDRE,cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur ecriture numeros de familles polyedres'
	   call efexit(-1)
	endif      
	print *,'Ecriture des numeros de familles des polyedres'

C       ** Fermeture du fichier                            **
	call efferm (fid,cret)
	print *,cret
	if (cret .ne. 0 ) then
	   print *,'Erreur fermeture du fichier'
	   call efexit(-1)
	endif      
	print *,'Fermeture du fichier'
C
	end
